packages/list/index.js

```diff
@@ -84,219 +87,199 @@ export default class List extends Component {
-        const listItem = this.listItems_[id];
-        if (listItem) {
-          listItem.toggleCheckbox();
-        }
+        this.setState({toggleCheckboxAtIndex: index});
       },
     };
   }
 
-  getListItemIdFromIndex_ = (index) => {
-    for (let id in this.listItemIndices_) {
-      if (this.listItemIndices_[id] === index) return id;
-    };
-    return '';
-  }
+  render() {
+    const {
+      /* eslint-disable no-unused-vars */
+      className,
+      nonInteractive,
+      dense,
+      avatarList,
+      twoLine,
+      singleSelection,
+      selectedIndex,
+      handleSelect,
+      wrapFocus,
+      /* eslint-enable no-unused-vars */
+      children,
+      ...otherProps
+    } = this.props;
 
-  getIndexOfListItemElement_ = (element) => {
-    let listItemId = '';
-    for (let id in this.listItems_) {
-      if (this.listItems_[id].listItemElement === element) {
-        listItemId = id;
-        break;
-      }
-    }
-    return this.listItemIndices_[listItemId];
+    this.listItemCount = 0;
+    return (
+      <ul
+        className={this.classes}
+        {...otherProps}
+      >
+        {React.Children.map(children, this.renderChild)}
+      </ul>
+    );
   }
 
-  getListItemIndexOfTarget_ = (eventTarget) => {
-    let target = eventTarget;
-
-    // Find the first ancestor that is a list item.
-    while (this.getIndexOfListItemElement_(target) === undefined) {
-      if (!target || target === document) {
-        return -1;
-      }
-      target = target.parentElement;
+  renderChild = (child) => {
+    if (child.type.name === 'ListItem') {
+      return this.renderListItem(child, this.listItemCount++);
+    } else {
+      return child;
     }
-
-    return this.getIndexOfListItemElement_(target);
   }
 
-  onKeyDown = (e) => {
-    this.props.onKeyDown(e);
+  handleKeyDown = (e, index) => {
     e.persist(); // Persist the synthetic event to access its `key`
-    const index = this.getListItemIndexOfTarget_(e.target);
     if (index >= 0) {
       this.foundation_.handleKeydown(e, true /* isRootListItem is true if index >= 0 */, index);
+      // Work around until MDC Web issue is resolved:
+      // https://github.com/material-components/material-components-web/issues/4053
+      if (e.key === 'Enter' || e.keyCode === 13 || e.key === 'Space' || e.keyCode === 32) {
+        this.props.handleSelect(index);
+      }
     }
   }
 
-  onClick = (e) => {
-    this.props.onClick(e);
-    const index = this.getListItemIndexOfTarget_(e.target);
+  handleClick = (e, index) => {
     // Toggle the checkbox only if it's not the target of the event, or the checkbox will have 2 change events.
     const toggleCheckbox = e.target.type === 'checkbox';
     this.foundation_.handleClick(index, toggleCheckbox);
+    // Work around until MDC Web issue is resolved:
+    // https://github.com/material-components/material-components-web/issues/4053
+    this.props.handleSelect(index);
```

Comment:
this should have an `if (index >= 0)` check. `foundation.handleClick` does that check, so we should also do it here.